### PR TITLE
update plugin documentation under docs>tasks>extend-kubectl

### PR DESCRIPTION
**Pull request by juanvallejo**
Updates `kubectl` plugin documentation under `docs > tasks > extend-kubectl` to include information and examples [relevant to the plugin mechanism update in 1.12](https://github.com/kubernetes/kubernetes/pull/66876).

cc @soltysh 

/sig cli

### CODE REVIEW

**Comment by juanvallejo** [comment]
@tengqm Thanks, review comment addressed

**Comment by juanvallejo** [comment]
@tengqm @mistyhacks friendly ping

**Comment by tengqm** [comment]
@juanvallejo This should go to release-1.12 branch?

**Comment by soltysh** [comment]
@tengqm that is correct

**Comment by juanvallejo** [comment]
/hold cancel

@soltysh @tengqm this PR should be good to go

File: content/en/docs/tasks/extend-kubectl/kubectl-plugins.md [hunk 1]
```diff
@@ -10,131 +11,266 @@ content_template: templates/task
...
+
+/usr/local/bin/plugins/kubectl-foo
+/usr/local/bin/moreplugins/kubectl-foo
+  - warning: /usr/local/bin/moreplugins/kubectl-foo is overshadowed by a similarly named plugin: /usr/local/bin/plugins/kubectl-foo
+
+error: one plugin warning was found
 ```
 
-### Accessing runtime attributes
+In the above scenario, the warning under `/usr/local/bin/moreplugins/kubectl-foo` tells us that this plugin will never be executed. Instead, the executable that appears first in our PATH, `/usr/local/bin/plugins/kubectl-foo`, willalways be found and executed first by the `kubectl` plugin mechanism.
```
Comment (kbhawkey) [inline content/en/docs/tasks/extend-kubectl/kubectl-plugins.md]: nit: will always

File: content/en/docs/tasks/extend-kubectl/kubectl-plugins.md [hunk 2]
```diff
@@ -10,131 +11,266 @@ content_template: templates/task
...
+A way to resolve this issue is to ensure that the location of the plugin that you wish to use with `kubectl` always comes first in your PATH. For example, if we wanted to always use `/usr/local/bin/moreplugins/kubectl-foo` anytime that the `kubectl` command `kubectl foo` was invoked, we would simply change the value of our PATH to be `PATH=/usr/local/bin/moreplugins:/usr/local/bin/plugins`.
 
-Plugins have access to runtime context attributes through environment variables. So to access the value provided through a flag, for example, just look for the value of the proper environment variable using the appropriate function call for your binary or script. 
+#### Invocation of the longest executable filename
 
-The supported environment variables are:
+There is another kind of overshadowing that can occur with plugin filenames. Given two plugins present in a user's PATH `kubectl-foo-bar` and `kubectl-foo-bar-baz`, the `kubectl` plugin mechanism will always choose the longest possible plugin name for a given user command. Some examples below, clarify this further:
```
Comment (kbhawkey) [inline content/en/docs/tasks/extend-kubectl/kubectl-plugins.md]: nit: Possibly make this a note?

**Comment by tfogo** [comment]
If this is for 1.12 it should not be against master.

/hold

We can see what @zparnold says but I think the best thing might be to hold off on this until the `release-1.12` branch hits master.

File: content/en/docs/tasks/extend-kubectl/kubectl-plugins.md [hunk 3]
```diff
@@ -10,131 +11,266 @@ content_template: templates/task
...
+of plugins as a means of utilizing these building blocks to create more complex behavior. Plugins extend `kubectl` with new sub-commands, allowing for new and custom features not included in the main distribution of `kubectl`.
 
 {{% /capture %}}
 
 {{% capture prerequisites %}}
 
 You need to have a working `kubectl` binary installed. 
 {{< note >}}
-**Note:** Plugins were officially introduced as an alpha feature in the v1.8.0 release. So, while some parts of the plugins feature were already available in previous versions, a `kubectl` version of 1.8.0 or later is recommended.
+**Note:** Plugins were officially introduced as an alpha feature in the v1.8.0 release. They have been re-worked in the v1.12.0 release to support a wider range of use-cases. So, while some parts of the plugins feature were already available in previous versions, a `kubectl` version of 1.12.0 or later is recommended if you are following these docs.
```
Comment (kbhawkey) [inline content/en/docs/tasks/extend-kubectl/kubectl-plugins.md]: nit: Possibly remove `So`, capitalize while.
Is it, While some parts of the plugins features were already available, or While some parts of the plugins feature was already available in previous versions...?

**Comment by tfogo** [comment]
@juanvallejo To add a bit of clarification here, the `release-1.12` branch is being merged today so we probably don't want to add something last minute. The deadline for additions to the 1.12 branch was a while ago and this PR should have been made against the `release-1.12` branch. Sorry if the process for getting these docs into the 1.12 release was unclear. But we can definitely work on getting them into the docs shortly after the initial release 👍

File: content/en/docs/tasks/extend-kubectl/kubectl-plugins.md [hunk 4]
```diff
@@ -10,131 +11,266 @@ content_template: templates/task
...
+I am a plugin with a dash in my name
+
+# it can also be inovked using an underscore
+$ kubectl foo_bar
+I am a plugin with a dash in my name
+```
+
+#### Name conflicts and overshadowing
+
+It can be possible to have multiple pluins with the same filename in different locations throughout your PATH.
```
Comment (kbhawkey) [inline content/en/docs/tasks/extend-kubectl/kubectl-plugins.md]: pluins -> plugins

File: content/en/docs/tasks/extend-kubectl/kubectl-plugins.md [hunk 5]
```diff
@@ -10,131 +11,266 @@ content_template: templates/task
...
-└── targaryen
-    ├── plugin.yaml
-    └── dracarys
+
+As you can see, our plugin was found based on the `kubectl` command specified by a user, and all extra arguments and flags were passed as-is to the plugin executable once it was found.
+
+#### Names with dashes and underscores
+
+Although the `kubectl` plugin mechanism uses the dashes (`-`) in plugin filenames to determine the sequence of sub-commands that should invoke them, it is still possible to create a plugin
+command containing dashes in its commandline invocation by using underscores `_` in its filename.
```
Comment (kbhawkey) [inline content/en/docs/tasks/extend-kubectl/kubectl-plugins.md]: nit: to be consistent, use (`_`)
nit, how does this read to you: uses the dash (`-`) in plugin filenames to separate the sequence of sub-commands processed by the plugin, it is still possible to create ...

**Pull request merged**